Mathlib/Topology/Compactness/KSpace.lean

```diff
@@ -0,0 +1,315 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn, Hannah Scholz
+-/
+
+import Mathlib.Topology.Defs.Induced
+import Mathlib.Topology.RestrictGen
+import Mathlib.Topology.Compactness.CompactlyGeneratedSpace
+
+/-!
+# Kspaces and the k-ification
+
+In this file we will define k-spaces and the k-ification and prove basic properties about them.
+This is a weaker version of `CompactlyGeneratedSpace`. These notions agree on Hausdorff spaces.
+They are both referred to as compactly generated spaces in the literature.
+
+## Main definitions
+* `KSpace`: A k-space is a topological space in which a set `A` is open iff for every compact set
+  `B`, the intersection `A ∩ B` is open in `B`.
+* `instkification`: For a topological space `X` one can define another topology on `X` as follows:
+  `A` is open iff for all compact sets `B`, the intersection `A ∩ B` is open in `B`.
+
+## Main results
+* `kspace_of_WeaklyLocallyCompactSpace`: every weakly locally compact space is a k-space
+* `kspace_of_SequentialSpace`: every sequential space is a k-space
+* `kspace_of_compactlyGeneratedSpace`: every compaclty generated space is a k-space.
+* `isCompact_iff_isCompact_in_kification`: The compact sets of a topological space and its
+  k-ification agree.
+* `kspace_kification`: The k-ification makes any space into a k-space.
+* `kification_kspace_eq_self`: The k-ification of a k-space `X` preserves the topology on `X`.
+* `continuous_kification_of_continuousOn_compact`: If a map `f : X → Y` is continuous on every
+  compact subset of `X` then it is continuous when viewed as a map from `kification X` to
+  `kification Y`.
+
+## References
+* [J. Munkres, *Topology*][Munkres2000]
+* <https://en.wikipedia.org/wiki/Compactly_generated_space>
+-/
+
+noncomputable section
+
+open Set Set.Notation Topology
+
+/-! ### K-spaces-/
+
+/-- A space is a KSpace if the topology is generated by the compact sets.-/
+class KSpace  (X : Type*) [TopologicalSpace X] : Prop where
+  /-- A space is a k-space if the topology is generated by the compact sets.-/
+  restrictGenTopology : RestrictGenTopology (X := X) {K | IsCompact K}
+
+namespace KSpace
+
+variable {X : Type*} [TopologicalSpace X] [KSpace X]
+
+/-- A set `A` in a k-space is open iff for every compact set `K`, the intersection `K ∩ A` is
+open in `K`.-/
+lemma isOpen_iff (A : Set X) : IsOpen A ↔ ∀ K, IsCompact K → IsOpen (K ↓∩ A) :=
+  RestrictGenTopology.isOpen_iff restrictGenTopology
+
+/-- A set `A` in a k-space is closed iff for every compact set `K`, the intersection `K ∩ A` is
+closed in `K`.-/
+lemma isClosed_iff (A : Set X) : IsClosed A ↔ ∀ K, IsCompact K → IsClosed (K ↓∩ A) :=
+  RestrictGenTopology.isClosed_iff restrictGenTopology
+
+/-- If every set `A` is open if for every compact `K` the intersection `K ∩ A` is open in `K`,
+then the space is a k-space. -/
+lemma kspace_of_isOpen {X : Type*} [TopologicalSpace X]
```

Comment:
Why not `of_isOpen`? The full name will be `KSpace.of_isOpen`, which IMHO makes sense. Also, please reuse `X` from `variable`s.